### DISCUSSION TITLE
firewallcmd-ipset native firewallcmd ipset commands

### DISCUSSION THREAD

**Discussion by ghost** [fail2ban/fail2ban#3817] [Ideas]
Hello!

Is there a specific reason the native firewallcmd commands are not used that allow showing the ipset stuff via firewallcmd?

`firewall-cmd --direct --new-ipset=<ipmset> --type=hash:ip --option=timeout=<default-ipsettime> <firewalld_familyopt>`
This works but doesn't allow firewall-cmd to show the content of the ipsets.

Is there a reason against using the native commands here?
```
sudo firewall-cmd --permanent --new-ipset=test-v4 --type=hash:ip --family=inet
sudo firewall-cmd --permanent --new-ipset=test-v6 --type=hash:ip --family=inet6
sudo firewall-cmd --get-entries --ipset=test-v4 --permanent
sudo firewall-cmd --permanent --ipset=test-v4 --add-entry=1.2.3.4
sudo firewall-cmd --permanent --ipset=test-v6 --add-entry=2001:db8::1
#restart of firewalld needed to apply new IPs
sudo firewall-cmd --permanent --zone=block --add-source=ipset:test-v4
sudo firewall-cmd --delete-ipset=test-v4 --permanent
#ipset must be removed from zone manually, instead firewalld can't restart
sudo firewall-cmd --permanent --zone=block --remove-source=ipset:test-v4
sudo firewall-cmd --get-ipsets --permanent
```

Are there better ways to handle it in a way that firewall-cmd shows the ipsets and their contents?

kind regards

**Comment by sebres** [comment]
> Are there better ways to handle it in a way that firewall-cmd shows the ipsets and their contents?

If I correct understand the discription - see #2620 and amends (in newer version both actions are combined to single action) 
and it is enough to specify `ipsetbackend=firewalld` for `firewallcmd-ipset`, see https://github.com/fail2ban/fail2ban/blob/54c0effceb998b73545073ac59c479d9d9bf19a4/config/action.d/firewallcmd-ipset.conf#L95-99

For example:
```ini
[jail_or_default]
banaction = firewallcmd-ipset[ipsetbackend=firewalld]
```

> Is there a specific reason the native firewallcmd commands are not used that allow showing the ipset stuff via firewallcmd?

Doesn't `ipset list` shows it well?

> Is there a reason against using the native commands here?

Backwards compatibility... 
But from other point of view I don't see any particular reason to use firewalld-based actions at all (why not native net-filters like iptables-ipset or nftables)... firewalld is just a wrapper over native net-filter and it's a bit strange to use it for banning.

**Comment by ghost** [comment]
Am 17.08.2024 um 07:15:01 Uhr schrieb Sergey G. Brester:

> > Are there better ways to handle it in a way that firewall-cmd shows
> > the ipsets and their contents?  
> 
> If I correct understand the discription - see #2620 and amends (in
> newer version both actions are combined to single action) and it is
> enough to specify `ipsetbackend=firewalld` for `firewallcmd-ipset`,
> see https://github.com/fail2ban/fail2ban/blob/54c0effceb998b73545073ac59c479d9d9bf19a4/config/action.d/firewallcmd-ipset.conf#L95-99
> 
> For example:
> ```ini
> [jail_or_default]
> banaction = firewallcmd-ipset[ipsetbackend=firewalld]
> ```

At my setup only ipsettype exists (the renaming isn't in this
version yet) and if I set it to firewalld, I get an error when banning:

ERROR   Failed to execute ban jail 'rslight' action 'firewallcmd-ipset' info 'ActionInfo({'ip': '172.17.1.5', 'family': 'inet4', 'fid': <function Actions.ActionInfo.<lambda> at 0x7f710757b560>, 'raw-ticket': <function Actions.ActionInfo.<lambda> at 0x7f710757bce0>})': Error starting action Jail('rslight')/firewallcmd-ipset: 'Script error'

Please tell me if you need further info from my machine.
Fail2Ban v1.1.0 on Debian Unstable.

> > Is there a specific reason the native firewallcmd commands are not
> > used that allow showing the ipset stuff via firewallcmd?  
> 
> Doesn't `ipset list` shows it well?

That lists it but then firewalld doesn't know about it.

> > Is there a reason against using the native commands here?  
> 
> Backwards compatibility... 
> But from other point of view I don't see any particular reason to use
> firewalld-based actions at all (why not native net-filters like
> iptables-ipset or nftables)... firewalld is just a wrapper over
> native net-filter and it's a bit strange to use it for banning.

Rather easy: One place to administer firewall stuff.

**Reply by sebres** [reply]
> Rather easy: One place to administer firewall stuff.

Why do you need to administer fail2ban stuff via firewalld at all?
